src/Core/Auth/Identity/EmailTokenProvider.cs

```diff
@@ -10,100 +9,55 @@ namespace Bit.Core.Auth.Identity;
 
 public class EmailTokenProvider : IUserTwoFactorTokenProvider<User>
 {
-    private const string CacheKeyFormat = "Email_TOTP_{0}_{1}";
+    private const string CacheKeyFormat = "EmailToken_{0}_{1}_{2}";
 
-    private readonly IServiceProvider _serviceProvider;
     private readonly IDistributedCache _distributedCache;
     private readonly DistributedCacheEntryOptions _distributedCacheEntryOptions;
 
     public EmailTokenProvider(
-        IServiceProvider serviceProvider,
         [FromKeyedServices("persistent")]
         IDistributedCache distributedCache)
     {
-        _serviceProvider = serviceProvider;
         _distributedCache = distributedCache;
         _distributedCacheEntryOptions = new DistributedCacheEntryOptions
         {
-            AbsoluteExpirationRelativeToNow = TimeSpan.FromMinutes(20)
+            AbsoluteExpirationRelativeToNow = TimeSpan.FromMinutes(5)
         };
     }
 
-    public async Task<bool> CanGenerateTwoFactorTokenAsync(UserManager<User> manager, User user)
-    {
-        var provider = user.GetTwoFactorProvider(TwoFactorProviderType.Email);
-        if (!HasProperMetaData(provider))
-        {
-            return false;
-        }
+    public int TokenLength { get; protected set; } = 8;
+    public bool TokenAlpha { get; protected set; } = false;
+    public bool TokenNumeric { get; protected set; } = true;
 
-        return await _serviceProvider.GetRequiredService<IUserService>().
-            TwoFactorProviderIsEnabledAsync(TwoFactorProviderType.Email, user);
+    public virtual Task<bool> CanGenerateTwoFactorTokenAsync(UserManager<User> manager, User user)
+    {
+        return Task.FromResult(!string.IsNullOrEmpty(user.Email));
     }
 
-    public Task<string> GenerateAsync(string purpose, UserManager<User> manager, User user)
+    public virtual async Task<string> GenerateAsync(string purpose, UserManager<User> manager, User user)
     {
-        var provider = user.GetTwoFactorProvider(TwoFactorProviderType.Email);
-        if (!HasProperMetaData(provider))
-        {
-            return null;
-        }
-
-        return Task.FromResult(RedactEmail((string)provider.MetaData["Email"]));
+        var code = CoreHelpers.SecureRandomString(TokenLength, TokenAlpha, true, false, TokenNumeric, false);
+        var cacheKey = string.Format(CacheKeyFormat, user.Id, user.SecurityStamp, purpose);
+        await _distributedCache.SetAsync(cacheKey, Encoding.UTF8.GetBytes(code), _distributedCacheEntryOptions);
+        return code;
     }
 
     public async Task<bool> ValidateAsync(string purpose, string token, UserManager<User> manager, User user)
     {
-        var cacheKey = string.Format(CacheKeyFormat, user.Id, token);
+        var cacheKey = string.Format(CacheKeyFormat, user.Id, user.SecurityStamp, purpose);
```

Comment:
Mirroring some of the implementation details from [ASP.NET Core's EmailTokenProvider](https://github.com/dotnet/aspnetcore/blob/main/src/Identity/Extensions.Core/src/EmailTokenProvider.cs), which includes security stamp as a parameter in the token generation. We can include it as part of the cache key to get similar benefits.